render/pixman/renderer.c

```diff
@@ -0,0 +1,339 @@
+	struct wlr_pixman_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static const struct wlr_texture_impl texture_impl;
+
+static struct wlr_pixman_texture *get_texture(
+		struct wlr_texture *wlr_texture) {
+	assert(wlr_texture->impl == &texture_impl);
+	return (struct wlr_pixman_texture *)wlr_texture;
+}
+
+static bool texture_is_opaque(struct wlr_texture *wlr_texture) {
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	return texture->format == WL_SHM_FORMAT_XRGB8888;
+}
+
+static bool texture_write_pixels(struct wlr_texture *wlr_texture,
+		uint32_t stride, uint32_t width, uint32_t height,
+		uint32_t src_x, uint32_t src_y, uint32_t dst_x, uint32_t dst_y,
+		const void *data) {
+	assert(false && "todo texture_write_pixels");
+	return false;
+}
+
+static bool texture_to_dmabuf(struct wlr_texture *wlr_texture,
+		struct wlr_dmabuf_attributes *attribs) {
+	assert(false && "todo texture_to_dmabuf");
+	return false;
+}
+
+static void texture_destroy(struct wlr_texture *wlr_texture) {
+	if (wlr_texture == NULL) {
+		return;
+	}
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	pixman_image_unref(texture->image);
+	free(texture);
+}
+
+static const struct wlr_texture_impl texture_impl = {
+	.is_opaque = texture_is_opaque,
+	.write_pixels = texture_write_pixels,
+	.to_dmabuf = texture_to_dmabuf,
+	.destroy = texture_destroy,
+};
+
+static uint32_t get_pixman_format_from_wl(enum wl_shm_format wl_fmt) {
+	uint32_t fmt = 0;
+	switch(wl_fmt) {
+	case WL_SHM_FORMAT_ARGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_a8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8a8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	case WL_SHM_FORMAT_XRGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_x8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8x8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	default:
+		// TODO: wl_shm_format to string?
+		wlr_log(WLR_ERROR, "Unsupported wl_shm_format");
+		break;
+	}
+
+	return fmt;
+}
+
+struct wlr_pixman_texture *pixman_create_texture(
+		struct wlr_texture *wlr_texture, struct wlr_pixman_renderer *renderer);
+
+static void handle_destroy_buffer(struct wl_listener *listener, void *data) {
+	struct wlr_pixman_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	if (!wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
```

Comment:
Other allocators may not support DMA-BUFs. It would be nice if the information we need was returned from `wlr_buffer_get_data_ptr`, or was part of the `wlr_buffer` struct.

Reply:
`wlr_buffer_get_format` ?

Reply:
We already have `wlr_buffer.width` and `wlr_buffer.height`. I think we could have `wlr_buffer.format` with a DRM format in it.

Another solution would be to add the DRM format as an out-param of `get_data_ptr`.

Reply:
I prefer having `wlr_buffer.format` and keep `get_data_ptr` for only getting the data.